### DISCUSSION TITLE
Sql-injections

### DISCUSSION THREAD

**Discussion by eydun** [CIDRAM/CIDRAM#508] [General]
Thanks for building this awesome library.

Does Cidram also block sql-injections attempts?

**Comment by Maikuolan** [comment]
> Thanks for building this awesome library.

You're welcome, and thank you for the compliment. :-)

> Does Cidram also block sql-injections attempts?

Yes, to a limited extent. As an example of some of the signatures CIDRAM has for blocking SQLi:

https://github.com/CIDRAM/CIDRAM/blob/9cce4bb3826355d15ba08f0fa9af5e2333fe4270/vault/modules/extras.php#L246-L250

If a particular pattern in requests where SQL injection has been found to have occurred is found to be reasonably common, and someone reports it, or we find some information related to that pattern, we'll likely add new signatures to the modules in order for CIDRAM to be aware of such requests. Many well-known *sources* of SQL injection are also blocked outright by CIDRAM's default signature files, and as a result, it's not unusual to see SQL injection attempts blocked by CIDRAM, not due to the attempt itself, but due to where the attempt is originating from.

However, it should be noted that not all websites are the same, many websites have unique structures, are built differently to other websites, and may, by their nature, be protected from vulnerabilities affecting many other websites, while being susceptible to numerous vulnerabilities of their own. An inherently insecure website not built with security in mind will never be absolutely secure, and while using CIDRAM will certainly help the situation, it still doesn't guarantee that the website won't ultimately be hacked and its database ultimately compromised at some point. Much the same is true for every other similar package and security solution out there: While they may all be helpful to varying degrees, using them isn't a substitute for ensuring the website is built securely.

Kind of like if you're running a nightclub, and you hire a well-reputed bouncer to guard the front door from unwanted guests. The bouncer may do a very good job at guarding that door, but if you've got a broken, unlockable window somewhere on the other side of the building and nobody is keeping an eye on it, and some of your unwanted guests are aware of it, there's always the possibility that they might be able to avoid your bouncer entirely by going through that window instead of the front door. Having the bouncer (i.e., CIDRAM and various security solutions) is still going to be useful, but ideally, you'll want to get that broken, unlockable window (i.e., the security vulnerabilities inherent in the website) bricked up, boarded up, or repaired as soon as possible, too.

Even blocking well-known sources of SQLi doesn't absolutely guarantee anything, as it's easy enough for a determined attacker to simply attack from a different source, by using a VPN, proxy, a different internet connection, or similar. (Though, blocking well-known sources of SQLi is still absolutely recommended all the same, as it'll still block the vast majority of such attacks, will help against many automated attacks, and against the less-determined, more lazy attackers out there, and logged block events could help to provide additional data).

Some useful reading material:
- *[What is SQL injection and how to prevent it?](https://docs.php.earth/security/sql-injection/)*
- *[PHP MySQLi Prepared Statements Tutorial to Prevent SQL Injection](https://websitebeaver.com/prepared-statements-in-php-mysqli-to-prevent-sql-injection)*
- *[SQL injection - a community paradoxon](https://www.tumblr.com/the-phlog/129182968120/sql-injection-a-community-paradoxon)*
- *[How to secure PHP web applications and prevent attacks?](https://docs.php.earth/security/intro/)*

In response to the above, some users may potentially ask at some point, "Why doesn't CIDRAM just escape everything itself, before any POST data, GET data, queries, etc, have the opportunity to reach the website or application in the first place?"

The main reason is that doing so would be tampering with that website's data, could potentially break functionality at that website (because again, not all websites are the same, and we can't always predict what the intended purpose may be for the POST/GET/query data received at that website, or whether messing around with that data would have unforeseen consequences). It's also possible, depending on what security measures may be implemented natively by the website, that data might already be being escaped, and attempting to escape it again could result in the data being escaped twice, effectively breaking the escaping and causing data to be unescaped again in some contexts, making the situation worse. For those websites which don't interact with SQL databases at all in the first place, or do, but make proper use of prepared statements, and implement other solutions natively which make escaping unnecessary, having everything be escaped might just be annoying for them, and might be something they would want to complain about. So, in short: It isn't something CIDRAM should have any business doing (and so doesn't).

But yeah. Using proper prepared statements and well-formed queries, or, when that isn't possible, properly escaping any inbound data which could potentially end up being used as part of an SQL query at some point, for the most part, should be sufficient. Always remember: Never implicitly trust any data originating from the client (e.g., inbound requests from users, browsers, bots, API calls, etc). :-)

**Comment by 737simpilot** [comment]
For me personally, I also use Ninjafirewall. But CIDRAM executes first then Ninja...
